python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1381 @@
+
+        for fragment in data_fragments:
+            fragments.push_back(fragment.unwrap())
+
+        simple_source = make_shared[CSimpleDataSource](fragments)
+        self.init(<shared_ptr[CDataSource]> simple_source)
+
+    cdef void init(self, const shared_ptr[CDataSource]& sp):
+        DataSource.init(self, sp)
+        self.simple_source = <CSimpleDataSource*> sp.get()
+
+
+cdef class TreeDataSource(DataSource):
+
+    cdef:
+        CTreeDataSource* tree_source
+
+    def __init__(self, data_sources):
+        cdef:
+            DataSource child
+            CDataSourceVector children
+            shared_ptr[CTreeDataSource] tree_source
+
+        for child in data_sources:
+            children.push_back(child.wrapped)
+
+        tree_source = make_shared[CTreeDataSource](children)
+        self.init(<shared_ptr[CDataSource]> tree_source)
+
+    cdef void init(self, const shared_ptr[CDataSource]& sp):
+        DataSource.init(self, sp)
+        self.tree_source = <CTreeDataSource*> sp.get()
+
+
+cdef class FileSystemDataSource(DataSource):
+
+    cdef:
+        CFileSystemDataSource* filesystem_source
+
+    def __init__(self, FileSystem filesystem not None, file_stats,
+                 Expression source_partition, dict path_partitions,
+                 FileFormat file_format not None):
+        cdef:
+            FileStats stats
+            Expression expression
+            vector[CFileStats] c_file_stats
+            shared_ptr[CExpression] c_source_partition
+            unordered_map[c_string, shared_ptr[CExpression]] c_path_partitions
+            CResult[shared_ptr[CDataSource]] result
+
+        for stats in file_stats:
+            c_file_stats.push_back(stats.unwrap())
+
+        for path, expression in path_partitions.items():
+            c_path_partitions[tobytes(path)] = expression.unwrap()
+
+        if source_partition is not None:
+            c_source_partition = source_partition.unwrap()
+
+        result = CFileSystemDataSource.Make(
+            filesystem.unwrap(),
+            c_file_stats,
+            c_source_partition,
+            c_path_partitions,
+            file_format.unwrap()
+        )
+        self.init(GetResultValue(result))
+
+    cdef void init(self, const shared_ptr[CDataSource]& sp):
+        DataSource.init(self, sp)
+        self.filesystem_source = <CFileSystemDataSource*> sp.get()
+
+
+cdef class Dataset:
+    """Collection of data fragments coming from possibly multiple sources."""
+
+    cdef:
+        shared_ptr[CDataset] wrapped
+        CDataset* dataset
+
+    def __init__(self, data_sources, Schema schema not None):
+        """Create a dataset
+
+        Parameters
+        ----------
+        data_sources : list of DataSource
+            One or more input data sources
+        schema : Schema
+            A known schema to conform to. The data sources must conform their
+            output to this schema with projections and filters taken into
+            account.
```

Comment:
It's not fully clear to me what this last sentence exactly means. 
Does it mean this scheme will be used to filter/project the output of the data sources, or that the other projections/filters specified when scanning should result in something that matches this schema?

Reply:
Well, it was not clear to me either, it is copied from the C++ apidocs. In case of projections this schema seems to be [omitted](https://github.com/apache/arrow/blob/8df0f5c13aa6692e2bdf31017627dfe5ba8159e3/cpp/src/arrow/dataset/scanner.cc#L133)?  cc @fsaintjacques